### PR TITLE
Implement offline run test

### PR DESCRIPTION
**Pull request by appukuttan-shailesh**
Implements #35 

**Approach:**
Process is split into 3 sub-tasks:

1. Prepare test for execution : `prep_run_test()`
2. Run the validation test : `run_test_offline()`
3. Register result on framework : `upload_test_result()`

Tasks 1 and 3 need internet access, while task 2 can be run offline.

`prep_run_test()` creates a JSON test config file. Example:
```
{
    "base_folder": "./hbp_validation_framework/02b7c22d-ee34-410d-9ce8-0d1ebfbb8bad/20181205-155501", 
    "test_observation_file": "./hbp_validation_framework/02b7c22d-ee34-410d-9ce8-0d1ebfbb8bad/20181205-155501/cell_density_Halasy_1996.json", 
    "params": {
        "param2": "b", 
        "param1": "a"
    }, 
    "test_instance_id": "54f3c567-6b6a-46e0-a214-e3adbfc3ca1b", 
    "test_id": "02b7c22d-ee34-410d-9ce8-0d1ebfbb8bad", 
    "test_instance_path": "morphounit.tests.CellDensityTest"
}
```
Also, the observation file is downloaded and stored locally.

`run_test_offline()` uses the above test config file and the observation file to create the test object. The test is run for the specified model and the results are stored locally as a pickled file.

Finally, `upload_test_result()` uploads the test result info onto the validation framework.

@apdavison : Does this seem fine? Also, I was wondering if we should update the existing `utils.run_test()` to reuse these functions so as to employ a single workflow for offline and online modes. I have created a temporary method called `run_test_new()` to demo this.

### CODE REVIEW

**Comment by appukuttan-shailesh** [comment]
@sasaray: Do let me know if you have any comments on this.

**Comment by appukuttan-shailesh** [comment]
Mechanism to avoid duplication can be incorporated after implementation of https://github.com/HumanBrainProject/hbp-validation-framework/issues/230

**Comment by appukuttan-shailesh** [comment]
@apdavison : Included the check for duplicates using hash values. Could you take a look and verify if things work fine?

**Comment by appukuttan-shailesh** [comment]
@apdavison : Shall we test this on the HPC resources next week?

File: hbp_validation_framework/utils.py [hunk 1]
```diff
@@ -8,13 +8,18 @@
 Action                                 Method
 ====================================   ====================================
 View JSON data in web browser          :meth:`view_json_tree`
+Prepare test for execution             :meth:`prep_run_test`
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: For clarity, I think this should be "prepare_run_test_offline"

File: hbp_validation_framework/utils.py [hunk 2]
```diff
@@ -8,13 +8,18 @@
...
+Run the validation test                :meth:`run_test_offline`
+Register result on framework           :meth:`upload_test_result`
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: "Register result with Validation Service"

File: hbp_validation_framework/utils.py [hunk 3]
```diff
@@ -70,11 +82,295 @@ def _make_js_file(data):
         json.dump(data, outfile)
         outfile.write("'")
 
-def run_test(hbp_username="", hbp_password=None, environment="production", model="", test_instance_id="", test_id="", test_alias="", test_version="", storage_collab_id="", register_result=True, client_obj=None, **test_kwargs):
+def prep_run_test(hbp_username="", hbp_password=None, environment="production", test_instance_id="", test_id="", test_alias="", test_version="", client_obj=None, **params):
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: For consistency with `__init__.py`, use `username` and `password`

File: hbp_validation_framework/utils.py [hunk 4]
```diff
@@ -70,11 +82,295 @@ def _make_js_file(data):
...
+    # score.exec_platform = str(self._get_platform())
+
+    # Save result info to file
+    test_result_file = os.path.join(test_info["base_folder"], "test_result.pkl")
+    with open(test_result_file, 'wb') as file:
+        pickle.dump(score, file)
+    return test_result_file
+
+def upload_test_result(hbp_username="", hbp_password=None, environment="production", test_result_file="", storage_collab_id="", register_result=True, client_obj=None):
+    """Register the result on the validation framework
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: Register the result with the Validation Service

File: hbp_validation_framework/utils.py [hunk 5]
```diff
@@ -70,11 +82,295 @@ def _make_js_file(data):
...
+
+    This method will register the validation result specified via the test result file
+    (generated by :meth:`run_test_offline`) on the validation framework.
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: with...service

File: hbp_validation_framework/utils.py [hunk 6]
```diff
@@ -70,11 +82,295 @@ def _make_js_file(data):
...
+    test_result_file = run_test_offline(model=model, test_config_file=test_config_file)
+    result_id, score = upload_test_result(hbp_username=hbp_username, hbp_password=hbp_password, environment=environment, test_result_file=test_result_file, storage_collab_id=storage_collab_id, register_result=register_result, client_obj=client_obj)
+    return result_id, score
+
+def run_test(hbp_username="", hbp_password=None, environment="production", model="", test_instance_id="", test_id="", test_alias="", test_version="", storage_collab_id="", register_result=True, client_obj=None, **params):
     """Run validation test and register result
 
     This method will accept a model, located locally, run the specified
-    test on the model, and store the results on the validation service.
+    test on the model, and store the results on the validation framework.
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: "service" is correct

File: hbp_validation_framework/utils.py [hunk 7]
```diff
@@ -173,7 +469,7 @@ def run_test(hbp_username="", hbp_password=None, environment="production", model
     print("----------------------------------------------")
 
     if register_result:
-        # Register the result with the HBP validation service
+        # Register the result with the HBP validation framework
```
Comment (apdavison) [inline hbp_validation_framework/utils.py]: service

File: hbp_validation_framework/utils.py (hunk 1 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

File: hbp_validation_framework/utils.py (hunk 2 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

File: hbp_validation_framework/utils.py (hunk 3 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

File: hbp_validation_framework/utils.py (hunk 4 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

File: hbp_validation_framework/utils.py (hunk 5 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

File: hbp_validation_framework/utils.py (hunk 6 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

File: hbp_validation_framework/utils.py (hunk 7 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/utils.py]: Done

**Comment by appukuttan-shailesh** [comment]
@apdavison : Changes made. We could make a new release (0.5?) once merged.

**Comment by apdavison** [comment]
Releasing 0.5.0 sounds good. Will you take care of that?

**Comment by appukuttan-shailesh** [comment]
Sure. Making the 0.5.0 release now.

**Pull request merged**